builtin/logical/pki/storage.go

```diff
@@ -76,6 +77,45 @@ func writeKey(ctx context.Context, s logical.Storage, key key) error {
 	return s.Put(ctx, json)
 }
 
+func resolveKeyReference(ctx context.Context, s logical.Storage, reference string) (keyId, error) {
+	if reference == "default" {
+		// Handle fetching the default key.
+		config, err := getKeysConfig(ctx, s)
+		if err != nil {
+			return keyId("config-error"), err
+		}
+
+		return keyId(config["default"]), nil
+	}
+
+	keys, err := listKeys(ctx, s)
+	if err != nil {
+		return keyId("list-error"), err
+	}
+
+	// Cheaper to list keys and check if an id is a match...
+	for _, key_id := range keys {
+		if key_id == keyId(reference) {
+			return key_id, nil
+		}
+	}
+
+	// ... than to pull all keys from storage.
+	for _, key_id := range keys {
+		key, err := fetchKeyById(ctx, s, key_id)
+		if err != nil {
+			return keyId("key-read"), err
+		}
+
+		if key.Name == reference {
+			return key.ID, nil
+		}
+	}
+
+	// Otherwise, we must not have found the key.
+	return keyId("not-found"), fmt.Errorf("unable to find PKI key for reference: %v", reference)
```

Comment:
This should probably be a UserError along with the one below?

Reply:
Possibly, again it's a convention thing where all the existing codebase within pki only checked for two error types in error handling code (InternalError and UserError). I *believe* I fixed all those handlers but there were a surprising number of them.